api/session.go

```diff
@@ -0,0 +1,341 @@
+package api
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/json"
+	"encoding/pem"
+	"fmt"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/canonical/lxd/lxd/response"
+	"github.com/canonical/lxd/shared"
+	"github.com/canonical/lxd/shared/logger"
+	"github.com/canonical/lxd/shared/trust"
+	"github.com/canonical/lxd/shared/ws"
+	"github.com/canonical/microcluster/rest"
+	"github.com/canonical/microcluster/state"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/canonical/microcloud/microcloud/api/types"
+	cloudClient "github.com/canonical/microcloud/microcloud/client"
+	cloudMDNS "github.com/canonical/microcloud/microcloud/mdns"
+	"github.com/canonical/microcloud/microcloud/service"
+)
+
+// HMACMicroCloud10 is the HMAC format version used during trust establishment.
+const HMACMicroCloud10 trust.HMACVersion = "MicroCloud-1.0"
+
+// SessionCmd represents the /1.0/session API on MicroCloud.
+var SessionCmd = func(sh *service.Handler) rest.Endpoint {
+	return rest.Endpoint{
+		AllowedBeforeInit: true,
+		Name:              "session",
+		Path:              "session",
+
+		Get: rest.EndpointAction{Handler: authHandlerMTLS(sh, sessionGet(sh))},
+	}
+}
+
+// sessionGet returns a MicroCloud join session.
+func sessionGet(sh *service.Handler) func(state *state.State, r *http.Request) response.Response {
+	return func(state *state.State, r *http.Request) response.Response {
+		if sh.ActiveSession() {
+			return response.BadRequest(fmt.Errorf("There already is an active session"))
+		}
+
+		sessionRole := r.URL.Query().Get("role")
+		if !shared.ValueInSlice(sessionRole, []string{"initiating", "joining"}) {
+			return response.BadRequest(fmt.Errorf("Invalid or missing session role"))
+		}
+
+		sessionTimeout, err := time.ParseDuration(r.URL.Query().Get("timeout"))
+		if err != nil {
+			return response.BadRequest(fmt.Errorf("Failed to parse timeout: %w", err))
+		}
+
+		return response.ManualResponse(func(w http.ResponseWriter) error {
+			conn, err := ws.Upgrader.Upgrade(w, r, nil)
+			if err != nil {
+				return err
+			}
+
+			defer func() {
+				err := conn.Close()
+				if err != nil {
+					logger.Error("Failed to close the websocket connection", logger.Ctx{"err": err})
+				}
+			}()
+
+			sessionCtx, cancel := context.WithTimeoutCause(r.Context(), sessionTimeout, fmt.Errorf("Session timeout exceeded"))
+			defer cancel()
+
```

Comment:
As it's just a query parameter I agree it would be nice to have a default. But if `timeout` has a default, should `role` also have one?

Maybe to overcome this, we can have a `/session/joining` and `/session/initiating` which optionally accepts a `?timeount` and defaults to `10m`. 
Then the `sessionGet` handler can accept both the `service.Handler` and the session type (joining, initiating) as argument.

Reply:
Implemented as discussed.